torchdata/datapipes/iter/load/huggingface.py

```diff
@@ -0,0 +1,66 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+import os
+import posixpath
+
+from typing import Any, Callable, Dict, Iterator, List, Optional, Tuple, Union
+
+from torch.utils.data.datapipes.utils.common import match_masks
+
+from torchdata.datapipes import functional_datapipe
+from torchdata.datapipes.iter import IterDataPipe
+from torchdata.datapipes.utils import StreamWrapper
+
+try:
+    import datasets
+except ImportError:
+    datasets = None
+
+def _get_response_from_huggingface_hub(dataset, split, revision, streaming, data_files) -> Tuple[Any, StreamWrapper]:
+    dataset = datasets.load_dataset(dataset, split, revision, streaming, data_files)
+    return dataset[0], StreamWrapper(dataset)
+
+class HuggingFaceHubReaderIterDataPipe(IterDataPipe[Tuple[str, StreamWrapper]]):
+    r"""
+    Takes in dataset names and returns an Iterable HuggingFace dataset
+    Args:
+        source_datapipe: a DataPipe that contains dataset names which will be accepted by the HuggingFace datasets library
+        revision: the specific dataset version
+        data_files: Optional dict to set custom train/test/validation split
+    Example:
+        >>> from torchdata.datapipes.iter import IterableWrapper, HuggingFaceHubReaderIterDataPipe
+        >>> huggingface_reader_dp = HuggingFaceHubReaderDataPipe(IterableWrapper(["lhoestq/demo1"]), revision="main")
+        >>> reader_dp = huggingface_reader_dp
+        >>> it = iter(reader_dp)
+        >>> path, line = next(it)
+        >>> path
+        Add test result here
+        >>> line
+        Add test result here b'BSD 3-Clause License'
+    """
+
+    source_datapipe: IterDataPipe[str]
+
+    def __init__(self, dataset: str, *, split : str = "train", revision : Optional[str] = None, streaming : bool = True, data_files : Optional[Dict[str,str]] = None) -> None:
+        if datasets is None:
+            raise ModuleNotFoundError(
+                "Package `datasets` is required to be installed to use this datapipe."
+                "Please use `pip install datasets` or `conda install -c conda-forge datasets`"
+                "to install the package"
+            )
+            
+        self.dataset = dataset
+        self.split = split
+        self.revision = revision
+        self.streaming = streaming
+        self.data_files = data_files
+
+    def __iter__(self) -> Iterator[Tuple[str, StreamWrapper]]:
+        yield _get_response_from_huggingface_hub(dataset=self.dataset, split=self.split, revision=self.revision, streaming=self.streaming, data_files=self.data_files)
+
+    def __len__(self) -> int:
+        return len([self.dataset])
```

Comment:
will `len(dp)` always be 1 in this case considering we only input a single `dataset`?

Reply:
yes `len(dp)` isn't particularly useful unfortunately here because the first input to `__init__` is a string

Reply:
how about returning the length of the HF [dataset](https://github.com/huggingface/datasets/blob/2f9a6a5bf3e25944b3620c870896dce0c8c0d52f/src/datasets/arrow_dataset.py#L1749-L1764)? However, it has different meaning compared to the one you implemented here. What do you think, @ejguan and @NivekT ~

Reply:
For the current implementation, I think the length as 1 makes more sense since only one element is yielded from it. You can treat it as an element with a large batch size = dataset size.

An alternative option is doing the similar thing as the [`TracingArrowDataPipe`](https://fburl.com/code/vpcqnqdp). Basically, we load the HF dataset as an internal object, and delegate operation like `map`, `shuffle`, etc. from datapipe to the internal HF dataset. And, when users iterate over the DataPipe instance, we will return the iterator of the internal HF dataset. In this case, we can return the length of internal HF Dataset.

cc: @VitalyFedyunin Not sure if this is the best practice for HF Dataset.